lib/ansible/plugins/inventory/docker_swarm.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018, Stefan Heitmueller <stefan.heitmueller@gmx.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: docker_swarm
+    plugin_type: inventory
+    authors:
+      - Stefan Heitmüller (stefan.heitmueller@gmx.com)
+    short_description: Ansible dynamic inventory plugin for Docker swarm nodes.
+    requirements:
+        - python >= 2.7
+        - Docker SDK for Python
+    description:
+        - Reads inventories from the Docker swarm API.
+        - Uses a YAML configuration file docker_swarm.[yml|yaml].
+    options:
+        plugin:
+            description: the name of this plugin, it should always be set to 'docker_swarm' for this plugin to recognize it as it's own.
+            required: true
+            choices: ['docker_swarm']
+        host:
+            description: Socket of a Docker swarm manager node (tcp,unix).
+            default: 'unix://var/run/docker.sock'
+'''
+
+EXAMPLES = '''
+# docker_swarm.yml
+plugin: docker_swarm
+host: unix:///var/run/docker.sock
+'''
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.release import __version__
+
+try:
+    import docker
+except ImportError:
+    raise AnsibleError('The Docker swarm dynamic inventory plugin requires the docker python binding: https://github.com/docker/docker-py.')
+
+
+class InventoryModule(BaseInventoryPlugin):
+    ''' Host inventory parser for ansible using Docker swarm as source. '''
+
+    NAME = "docker_swarm"
+
+    def _populate(self):
+        client = docker.DockerClient(base_url=self.get_option('host'))
+        self.inventory.add_group('manager')
+        self.inventory.add_group('worker')
+        try:
+            manager_nodes = client.nodes.list(filters={'role': 'manager'})
+            worker_nodes = client.nodes.list(filters={'role': 'worker'})
+        except ConnectionError:
+            raise AnsibleError('Unable to connect to the Docker swarm API.')
+        for node in manager_nodes:
+            node_hostname = client.nodes.get(node.id).attrs['Description']['Hostname']
```

Comment:
I think if this is going to be a dynamic inventory to use an ID instead of hostname. In Swarm, if a node leaves a swarm it is not deleted by the manager by default. When it rejoins you will have two entries with the same hostname - one will be marked as unreachable but still part of a swarm, the second one as active. You can also have multiple nodes with the same hostname and there is nothing wrong with it. It will be safer to provide an `ID` as an inventory entry for swarm modules so dynamic inventory covers all detected modules.